inc/ustring.bi

```diff
@@ -0,0 +1,681 @@
+
+PRIVATE CONSTRUCTOR DWSTR (BYREF cws AS DWSTR)
+  IF cws.m_BufferLen THEN
+     this.Add(cws)                                    'Add the passed DWSTR
+  ELSE
+     this.ResizeBuffer(m_GrowSize)                    'Create the initial buffer
+  END IF
+END CONSTRUCTOR
+
+PRIVATE CONSTRUCTOR DWSTR (BYVAL n AS LONGINT)
+  DIM wsz AS WSTRING * 260 = .WSTR(n)
+  this.Add(wsz)
+END CONSTRUCTOR
+
+PRIVATE CONSTRUCTOR DWSTR (BYVAL n AS DOUBLE)
+  DIM wsz AS WSTRING * 260 = .WSTR(n)
+  this.Add(wsz)
+END CONSTRUCTOR
+
+
+'***********************************************************************************************
+' Destructor
+'***********************************************************************************************
+
+
+PRIVATE DESTRUCTOR DWSTR
+  IF m_pBuffer THEN Deallocate(m_pBuffer)
+  m_pBuffer = 0
+END DESTRUCTOR
+
+
+'***********************************************************************************************
+' operators
+'***********************************************************************************************
+
+
+'***********************************************************************************************
+' One * returns the address of the DWSTR buffer.
+' Two ** deferences the string data.
+' We may use **cws (notice the double indirection) with these functions.
+'***********************************************************************************************
+
+
+PRIVATE OPERATOR * (BYREF cws AS DWSTR) AS WSTRING PTR
+  OPERATOR = cast(WSTRING PTR, cws.m_pBuffer)
+END OPERATOR
+
+
+'***********************************************************************************************
+
+
+PRIVATE OPERATOR LEN (BYREF cws AS DWSTR) AS ulong    'returns the length, in characters, of the DWSTR.
+  OPERATOR = cws.m_BufferLen \ cws.u_size
+END OPERATOR
+
+PRIVATE OPERATOR DWSTR.CAST () AS ANY PTR             'returns a pointer to the DWSTR buffer.
+  OPERATOR = cast(ANY PTR, m_pBuffer)
+END OPERATOR
+
+PRIVATE OPERATOR DWSTR.CAST () BYREF AS WSTRING       'returns the string data (same as **).
+  OPERATOR = *cast(WSTRING PTR, m_pBuffer)
+END OPERATOR
+
+
+#if sizeof(wstring) = 2
+PRIVATE OPERATOR DWSTR.[] (BYVAL nIndex AS ulong) byref AS USHORT
+'***********************************************************************************************
+' Returns the corresponding ASCII or Unicode integer representation of the character at the
+' zerobased position specified by the nIndex parameter. Can be used to change a value too.
+'***********************************************************************************************
+static zero as ushort                                 'fallback for nIndex outside valid data
+
+  IF nIndex > (m_BufferLen \ u_size) THEN
+    zero = &HFFFF
+    OPERATOR = zero                                   'return 0
+    exit operator
+  end if
+  
+  OPERATOR = *cast(USHORT ptr, m_pBuffer + (nIndex * u_size))
+END OPERATOR
+
+
+#else
+PRIVATE OPERATOR DWSTR.[] (BYVAL nIndex AS ulong) byref AS Ulong
```

Comment:
Feels like this conditional compilation could be combined with the one above, possibly using a `uchar` type.